components/org.wso2.carbon.identity.conditional.auth.functions.user/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/user/SetAccountAssociationToLocalUserImpl.java

```diff
@@ -33,29 +44,75 @@
     public void doAssociationWithLocalUser(JsAuthenticatedUser federatedUser, String username, String tenantDomain,
                                            String userStoreDomainName) {
 
-        if (federatedUser != null) {
-            if (federatedUser.getWrapped().isFederatedUser()) {
-                String externalSubject = federatedUser.getWrapped().getAuthenticatedSubjectIdentifier();
-                String externalIdpName = federatedUser.getWrapped().getFederatedIdPName();
-                if (externalSubject != null && externalIdpName != null) {
-                    associateID(externalIdpName, externalSubject, username, tenantDomain, userStoreDomainName);
+        String federatedIdpName = null;
+        try {
+            if (federatedUser != null) {
+                if (federatedUser.getWrapped().isFederatedUser()) {
+                    federatedIdpName = federatedUser.getWrapped().getFederatedIdPName();
+                    String userIdClaimURI = getUserIdClaimURI(federatedIdpName);
+                    String externalSubject;
+                    if (StringUtils.isNotEmpty(userIdClaimURI)) {
+                        externalSubject = federatedUser.getWrapped().getUserAttributes().entrySet().stream().filter(
+                                userAttribute -> userAttribute.getKey().getRemoteClaim().getClaimUri()
+                                        .equals(userIdClaimURI))
+                                .map(Map.Entry::getValue)
+                                .findFirst()
+                                .orElse(null);
+                    } else {
+                        externalSubject = federatedUser.getWrapped().getAuthenticatedSubjectIdentifier();
+                    }
+                    String externalIdpName = federatedUser.getWrapped().getFederatedIdPName();
+                    if (externalSubject != null && externalIdpName != null) {
+                        associateID(externalIdpName, externalSubject, username, tenantDomain, userStoreDomainName);
+                    } else {
+                        if (log.isDebugEnabled()) {
+                            log.debug(" Authenticated user or External IDP may be null " + " Authenticated User: " +
```

Comment:
I feel like better to add these as warning or error logs. As the association will not take place in these instances and caller will not have any idea.

Reply:
Same for other debugs as well.

Reply:
fixed in 4d66091